Various improvements related to CFLAGS

real-or-random opened the pull request:
- Fixes strange warnings in mingw builds (one item in #923).
 - Enables more and more fine-grained compiler warnings.
 - Fixes minor build issue on macOS  #896 on macOS.
 - Makes compiler warnings into errors on CI (`-Werror`, mentioned in https://github.com/bitcoin-core/secp256k1/issues/929#issuecomment-830861253).

real-or-random commented:
CI seems to be confused here because this is the same source branch as on #938 . Trying to close and reopen the PR to fix it.

real-or-random commented:
rebased

roconnor-blockstream commented:
LGTM. I did test building in under Nixos, but that was the extent of my testing.

hebasto commented on build-aux/m4/bitcoin_secp.m4:
```diff
@@ -82,3 +82,19 @@ if test x"$has_valgrind" != x"yes"; then
 fi
 ])
+
+dnl SECP_TRY_APPEND_CFLAGS([flags], VAR)
```
From my understanding, the square brackets `[...]` in a comment, that describes a new macro, means that parameter is optional. They do not mean quoting. Right?

real-or-random replied:
fixed

hebasto commented on configure.ac:
```diff
@@ -70,35 +66,40 @@ case $host_os in
    ;;
 esac
 
-CFLAGS="-W $CFLAGS"
```
Maybe document the removing of `-W` option?

real-or-random replied:
Apparently I've documented it here already https://github.com/bitcoin-core/secp256k1/pull/944/files#diff-49473dca262eeab3b4a43002adb08b4db31020d190caaad1594b47f1d5daa810R95

hebasto replied:
Ah, right.

hebasto reviewed (commented):
Approach ACK 64ec678daaf215e007d44433cdc9b969b6e09098, tested on:
- Linux Mint 20.1 (x86_64)
- macOS 11.4 (20F71)

Warnings [introduced](https://github.com/bitcoin-core/secp256k1/pull/700#discussion_r577840652) in #700 (reported in #896) are gone.

The last two commits seems the one to me, maybe squash them?
But maybe another approach could be considered?

Why a CI is not a "user" from the Automake's point of view? If so, why to introduce the `WERROR_CFLAGS` variable, while we have a dedicated `CFLAGS` one?

real-or-random commented:
> Why a CI is not a "user" from the Automake's point of view? If so, why to introduce the `WERROR_CFLAGS` variable, while we have a dedicated `CFLAGS` one?

It may seem overly complicated but I think it's better for testing/CI because it does not interfere with autoconf's default for `CFLAGS` (which is `-g -O2` or just `-g` depending on the compiler). What we typically want to test in CI is building with unset `CFLAGS`. We could always set `-Werror -g -O2` instead of `-Werror` but that seems complicated too.

By the way, I think I overlooked this line: https://github.com/bitcoin-core/secp256k1/pull/944/files#diff-49473dca262eeab3b4a43002adb08b4db31020d190caaad1594b47f1d5daa810R397

Will address the other comments.

hebasto commented:
> > Why a CI is not a "user" from the Automake's point of view? If so, why to introduce the `WERROR_CFLAGS` variable, while we have a dedicated `CFLAGS` one?
> 
> It may seem overly complicated but I think it's better for testing/CI because it does not interfere with autoconf's default for `CFLAGS` (which is `-g -O2` or just `-g` depending on the compiler). What we typically want to test in CI is building with unset `CFLAGS`. We could always set `-Werror -g -O2` instead of `-Werror` but that seems complicated too.

Ah, indeed. This repo does not have the `--enable-werror` configure option as Bitcoin Core has :)

real-or-random commented:
Oh interesting, maybe I should just have looked at Core then... ;) 

For context:
https://github.com/bitcoin-core/secp256k1/pull/938#issuecomment-839632758

Anyway, this thing here works and I don't want to touch it further unless necessary.

real-or-random commented:
Addressed the comments. I also rebased on master (sorry, that wasn't even intentional...)

> The last two commits seems the one to me, maybe squash them?

Done.

hebasto commented:
ACK a5c6ff2917edf97e888c441eb545ac70083b5376

jonasnick commented on configure.ac:
```diff
@@ -70,35 +66,40 @@ case $host_os in
+    SECP_TRY_APPEND_CFLAGS([-Wall], $1) # GCC >= 2.95 and probably many other compilers
+    SECP_TRY_APPEND_CFLAGS([-Wno-unused-function], $1) # GCC >= 3.0, -Wunused-function is implied by -Wall.
+    SECP_TRY_APPEND_CFLAGS([-Wextra], $1) # GCC >= 3.4, this is the newer name of -W, which we don't use because older GCCs will warn about unused functions.
+    SECP_TRY_APPEND_CFLAGS([-Wcast-align=strict], $1) # GCC >= 8.0
```
clang doesn't have `-Wcast-align=strict` but only `-Wcast-align`. Since this PR replaces the latter with the former, when using clang there are no alignment warnings anymore.

real-or-random commented:
fixed

jonasnick reviewed (commented):
ACK 0302138f7508414e9e5212bc45b4ca4c0e5f081c
